pysmt/sorting_networks.py

```diff
@@ -0,0 +1,333 @@
+#
+# This file is part of pySMT.
+#
+#   Copyright 2014 Andrea Micheli and Marco Gario
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+"""
+This module defines SortingNetwork class and functions
+"""
+
+import pysmt.environment
+
+
+class SortingNetwork(object):
+    """Provides an efficient encoding for counting elements. """
+
+    def __init__(self, mgr=None):
+        if mgr is not None:
+            self.mgr = mgr
+            self.env = self.mgr.env
+        else:
+            self.env = pysmt.environment.get_env()
+            self.mgr = self.env.formula_manager
+        self.And = self.mgr.And
+        self.Or = self.mgr.Or
+        self._cache = {}
+
+    def at_most_k(self, k, formulae):
+        assert(k > 0)
+        if(k >= len(formulae)):
+            return self.mgr.TRUE()
+
+        sn = self.sorting_network(formulae)
+
+        return self.mgr.Not(sn[k])
+
+    def at_least_k(self, k, formulae):
+        assert(k > 0)
+        if(k > len(formulae)):
+            return self.mgr.FALSE()
+
+        sn = self.sorting_network(formulae)
+        return sn[k-1]
+
+    def exactly_k(self, k, formulae):
+        assert(k > 0)
+        if(k > len(formulae)):
+            return self.mgr.FALSE()
+
+        sn = self.sorting_network(formulae)
+
+        if(k == len(formulae)):
+            return sn[k-1]
+        else:
+            return self.mgr.And(sn[k-1],
+                                self.mgr.Not(sn[k]))
+
+    def sorting_network(self, formulae):
+        """ Returns a list of sorting networks over formulae
+
+        Given formulae = phi1, phi2, ..., phiN, and defining the operator
+        true_evals(formulae) that returns the number of functions that evaluates to true.
+        The result of sorting_network(formulae) = gamma1, gamma2, ..., gammaN where
+        gamma1 <-> (true_evals(formulae) >= 1),
+        gamma2 <-> (true_evals(formulae) >= 2),
+        ...
+        gammaN <-> (true_evals(formulae) >= N),
+        """
+        key = tuple(formulae)
+        return self._cache.setdefault(key, self._sorting_network(formulae))
```

Comment:
You should check if ```key``` is in ```self._cache``` first, otherwise you uselessly compute the sorting network in any case!

(The actual parameters are computed before calling functions...)